nav2_controller/src/controller_server.cpp

```diff
@@ -284,6 +284,7 @@ ControllerServer::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
    */
   if (costmap_ros_->get_current_state().id() ==
     lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE)
+  //this condition could be removed
```

Comment:
We test it (removing these four conditions ) by repeating  `ros2 launch` and _`pre-shutdown` at random moment_ more than 1000 times in our local environment, as a result there's no shutdown error happening.

It now appears that the unit test of CI considers Costmap as an individual node for launch testing (launch it independntly). 

If we could change this unit test into relying on a virtual lifecycle node for launch testing, it would be more fitable for our suggestion.
After that, I believe that the coverage report may help us to prove these four conditions will be 100% judged correct.

Reply:
another way could also be tried:

change code into:
```cpp
  if (costmap_ros_->get_current_state().id() ==
    lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE)
{
.....
}
else {
    RCLCPP_INFO(get_logger(), “??????????????????");
}
```
after test, code line `RCLCPP_INFO(get_logger(), “??????????????????");` could not be covered anymore.

Reply:
Did you read the discussion to understand why those were there and its removal doesn't have any downsides after making this change?

>  It now appears that the unit test of CI considers Costmap as an individual node for launch testing (launch it independently). If we could change such unit test into relying on a virtual parent-LifecycleNode for launch and shutdown testing, it would be more fitable for our suggestion.

I don't understand. Are you saying that this doesn't work when the costmap nodes are launched independently? See other comment below, that needs to work as well.

Reply:
>Did you read the discussion to understand why those were there and its removal doesn't have any downsides after making this change?

of course, our solution could solve it together. unclean shutdown issue mentioned in #3377 is very similar to #3958 . 

we believe they're all caused by:
>all LifecycleNode react to rcl_preshutdown randomly,
&&
>costmap may react to ctrl+C earlier than its parents to cause nullptr-accessed and then, unclean shutdown.

so what we achieved is to make sure that `child-LifecycleNode`'s reaction to `rcl_preshutdown` is all up to its `parent-LifecycleNode`.


what's more, as #3377 said:
>There are a few instances fixed in this PR, but likely not all (see Future work below)

that's because `double check` method used before may be just a temporary solution. it's too hard for users to check all details and add `double check` in their whole programs.

so, we believe our solution is a thorough solution to all of similar issues in nav2-programs, and more friendly for users to use it.
they don't need to check too much and they don't need to add `double check` anymore. 
What they only need to do is to run `costmap_ros_->turnChildLifecycleNode()` following they set `  costmap_thread_ = std::make_unique<nav2_util::NodeThread>(costmap_ros_);`
doing so, `costmap_ros_`'s state will be completely up to its parents-LifecycleNode planner/controller_server.

Reply:
>Did you read the discussion to understand why those were there and its removal doesn't have any downsides after making this change?

of course, our solution could solve it together. unclean shutdown issue mentioned in #3377 is very similar to #3958 . 

they're all caused by:
>all LifecycleNode react to rcl_preshutdown randomly,
&&
>costmap may react to ctrl+C earlier than its parents to cause nullptr-accessed and then, unclean shutdown.

so what we achieved is to make sure that `child-LifecycleNode`'s reaction to `rcl_preshutdown` is all up to its `parent-LifecycleNode`.


what's more, as #3377 said:
>There are a few instances fixed in this PR, but likely not all (see Future work below)

that's because `double check` method used before may be just a temporary solution. it's too hard for users to check all details and add `double check` in their whole programs.

so, we believe our solution is a thorough solution to all of similar issues in nav2-programs, and more friendly for users to use it.
they don't need to check too many details and they don't need to add `double check` anymore. 
What they only need to do is to run `costmap_ros_->turnChildLifecycleNode()` following they set `  costmap_thread_ = std::make_unique<nav2_util::NodeThread>(costmap_ros_);`
doing so, `costmap_ros_`'s state will be completely up to its parents-LifecycleNode planner/controller_server.

Reply:
>What they only need to do is to run costmap_ros_->turnChildLifecycleNode() following they set   costmap_thread_ = std::make_unique<nav2_util::NodeThread>(costmap_ros_);

the case is like following:
```cpp
  // Launch a thread to run the costmap node
  costmap_thread_ = std::make_unique<nav2_util::NodeThread>(costmap_ros_);
  // work as a child-LifecycleNode of its parent planner_server
  costmap_ros_->turnChildLifecycleNode();
```

Reply:
> of course, our solution can cover the role of these double-checks.

OK, you can remove them - don't comment them out.